File: app/src/main/java/com/phoneapprove/app/data/SecureChannel.kt
package com.phoneapprove.app.data

import java.io.BufferedReader
import java.io.OutputStream
import java.math.BigInteger
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.PrivateKey
import java.security.PublicKey
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

class HandshakeException(message: String) : Exception(message)

/**
 * Kotlin mirror of daemon/secure_channel.py - has to match it byte-for-byte
 * (curve, public-key point encoding, HKDF construction, nonce scheme) or
 * the two sides derive different keys and every message fails to decrypt.
 * See that file for the full rationale: TCP has no equivalent of
 * Bluetooth's physical-proximity barrier, so this link needs its own
 * encryption now that anyone on the LAN can otherwise sniff it in the
 * clear. Built entirely on java.security/javax.crypto - no extra
 * dependency needed here, unlike the daemon's `cryptography` package.
 */
class SecureChannel private constructor(
    private val reader: BufferedReader,
    private val output: OutputStream,
    private val sendKey: ByteArray,
    private val recvKey: ByteArray,
) {
    private var sendCounter = 0L
    private var recvCounter = 0L
    private val sendLock = Any()

    fun sendLine(data: ByteArray) {
        // Locked end-to-end: concurrent callers interleaving encrypt+write
        // would both scramble the line framing and use nonces out of order.
        synchronized(sendLock) {
            val nonce = counterNonce(sendCounter)
            sendCounter++
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(sendKey, "AES"), GCMParameterSpec(128, nonce))
            val ciphertext = cipher.doFinal(data)
            output.write((bytesToHex(ciphertext) + "\n").toByteArray(Charsets.US_ASCII))
            output.flush()
        }
    }

    /** Returns decrypted bytes, or null on clean EOF. IOExceptions (including
     * a read timeout) propagate rather than collapsing to null, matching
     * the daemon side's recv_line contract. */
    fun recvLine(): ByteArray? {
        val line = reader.readLine() ?: return null
        val nonce = counterNonce(recvCounter)
        recvCounter++
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        return try {
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(recvKey, "AES"), GCMParameterSpec(128, nonce))
            cipher.doFinal(hexToBytes(line))
        } catch (e: Exception) {
            throw HandshakeException("decrypt failed (wrong token, or tampering): $e")
        }
    }

    companion object {
        private const val CURVE_NAME = "secp256r1"

        /** Client speaks first: send our public key, then read the peer's. */
        fun clientHandshake(reader: BufferedReader, output: OutputStream, token: String): SecureChannel {
            val (priv, pub) = generateKeyPair()
            writeKexLine(output, pub)
            val peerPub = readKexLine(reader)
            val shared = computeSharedSecret(priv, peerPub)
            val (c2s, s2c) = deriveKeys(shared, token)
            return SecureChannel(reader, output, sendKey = c2s, recvKey = s2c)
        }

        /** Server speaks second: read the peer's public key, then reply. */
        fun serverHandshake(reader: BufferedReader, output: OutputStream, token: String): SecureChannel {
            val (priv, pub) = generateKeyPair()
            val peerPub = readKexLine(reader)
            writeKexLine(output, pub)
            val shared = computeSharedSecret(priv, peerPub)
            val (c2s, s2c) = deriveKeys(shared, token)
            return SecureChannel(reader, output, sendKey = s2c, recvKey = c2s)
        }

        private fun generateKeyPair(): Pair<PrivateKey, PublicKey> {
            val kpg = KeyPairGenerator.getInstance("EC")
            kpg.initialize(ECGenParameterSpec(CURVE_NAME))
            val kp = kpg.generateKeyPair()
            return kp.private to kp.public
        }

        private fun ecParams(): ECParameterSpec {
            val params = AlgorithmParameters.getInstance("EC")
            params.init(ECGenParameterSpec(CURVE_NAME))
            return params.getParameterSpec(ECParameterSpec::class.java)
        }

        /** Raw uncompressed SEC1 point: 0x04 || X(32 bytes) || Y(32 bytes) -
         * matches what Python's cryptography library produces/expects via
         * Encoding.X962 + PublicFormat.UncompressedPoint. */
        private fun encodePoint(pub: PublicKey): ByteArray {
            val ecPub = pub as ECPublicKey
            val x = fixedLength(ecPub.w.affineX, 32)
            val y = fixedLength(ecPub.w.affineY, 32)
            return byteArrayOf(0x04) + x + y
        }

        private fun decodePoint(bytes: ByteArray): PublicKey {
            if (bytes.size != 65 || bytes[0] != 0x04.toByte()) {
                throw HandshakeException("bad EC point encoding (${bytes.size} bytes)")
            }
            val x = BigInteger(1, bytes.copyOfRange(1, 33))
            val y = BigInteger(1, bytes.copyOfRange(33, 65))
            val spec = ECPublicKeySpec(ECPoint(x, y), ecParams())
            return KeyFactory.getInstance("EC").generatePublic(spec)
        }

        /** ECDH output is the raw X-coordinate of the shared point, per SEC1/
         * JCA convention - matches Python's exchange() output exactly. */
        private fun computeSharedSecret(priv: PrivateKey, peerPub: PublicKey): ByteArray {
            val ka = KeyAgreement.getInstance("ECDH")
            ka.init(priv)
            ka.doPhase(peerPub, true)
            return fixedLength(BigInteger(1, ka.generateSecret()), 32)
        }

        /** RFC 5869 HKDF, collapsed to the one-block case (our output is
         * exactly 32 bytes = SHA-256's size, so Expand is a single HMAC
         * call): PRK = HMAC-SHA256(key=token, msg=shared_secret); each key
         * is HMAC-SHA256(PRK, info || 0x01). Must match secure_channel.py's
         * _derive_keys exactly. */
        private fun deriveKeys(sharedSecret: ByteArray, token: String): Pair<ByteArray, ByteArray> {
            val prk = hmacSha256(token.toByteArray(Charsets.UTF_8), sharedSecret)
            val c2s = hmacSha256(prk, "c2s".toByteArray(Charsets.UTF_8) + byteArrayOf(1))
            val s2c = hmacSha256(prk, "s2c".toByteArray(Charsets.UTF_8) + byteArrayOf(1))
            return c2s to s2c
        }

        private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
            val mac = Mac.getInstance("HmacSHA256")
            mac.init(SecretKeySpec(key, "HmacSHA256"))
            return mac.doFinal(data)
        }

        /** Unsigned big-endian encoding of `value`, exactly `length` bytes -
         * BigInteger.toByteArray() is two's-complement and can carry a
         * leading zero sign byte (or, for a value with fewer significant
         * bytes, be shorter), so strip/pad accordingly. */
        private fun fixedLength(value: BigInteger, length: Int): ByteArray {
            val raw = value.toByteArray()
            val trimmed = if (raw.size > length) raw.copyOfRange(raw.size - length, raw.size) else raw
            return if (trimmed.size < length) ByteArray(length - trimmed.size) + trimmed else trimmed
        }

        private fun writeKexLine(output: OutputStream, pub: PublicKey) {
            val json = "{\"type\":\"kex\",\"pub\":\"${bytesToHex(encodePoint(pub))}\"}\n"
            output.write(json.toByteArray(Charsets.UTF_8))
            output.flush()
        }

        private fun readKexLine(reader: BufferedReader): PublicKey {
            val line = reader.readLine() ?: throw HandshakeException("connection closed during key exchange")
            val hex = Regex("\"pub\"\\s*:\\s*\"([0-9a-fA-F]+)\"").find(line)?.groupValues?.get(1)
                ?: throw HandshakeException("expected kex message, got: $line")
            return decodePoint(hexToBytes(hex))
        }

        private fun counterNonce(counter: Long): ByteArray {
            val nonce = ByteArray(12)
            for (i in 0 until 8) {
                nonce[11 - i] = ((counter shr (8 * i)) and 0xFF).toByte()
            }
            return nonce
        }

        private fun bytesToHex(bytes: ByteArray): String = bytes.joinToString("") { "%02x".format(it) }

        private fun hexToBytes(hex: String): ByteArray = ByteArray(hex.length / 2) { i ->
            ((Character.digit(hex[i * 2], 16) shl 4) + Character.digit(hex[i * 2 + 1], 16)).toByte()
        }
    }
}

File: daemon/secure_channel.py
"""Wraps a raw stream socket with a per-connection encrypted channel:
ephemeral P-256 ECDH key exchange, HKDF-SHA256 key derivation with the
pre-shared pairing token mixed in, and AES-256-GCM for everything after.

Why this exists: the daemon<->phone link used to be Bluetooth RFCOMM, where
physical proximity was the main barrier to eavesdropping. Switching to plain
TCP over the local network removed that - anyone on the same LAN can now see
raw packets, and previously that meant they'd see every tool_name/tool_input/
cwd/reply in the clear, plus the pre-shared token itself in the "hello"
message. This makes that traffic unreadable to a passive eavesdropper and
resistant to an active MITM.

Security note on the MITM resistance specifically: the pairing token is
never sent over the network (it only ever travels out-of-band, via the QR
code/pairing text). Mixing it into the key derivation means an attacker who
intercepts and substitutes the ECDH public keys ends up deriving different
session keys than the two real endpoints do - so a tampered handshake just
produces AES-GCM authentication failures on the first real message, not a
silently-compromised channel.

P-256 (not X25519) specifically because it's natively supported by both
this library (`cryptography`) and Android's built-in java.security/
javax.crypto - no extra dependency needed on the Android side.
"""
import json
import struct
import threading

from cryptography.hazmat.primitives import hashes, hmac, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

_CURVE = ec.SECP256R1()
_KEX_TIMEOUT = 15


class HandshakeError(Exception):
    pass


def _hkdf_one_block(prk: bytes, info: bytes) -> bytes:
    """RFC 5869 HKDF-Expand for exactly one 32-byte block (all we need,
    since our output length equals SHA-256's block size) - T(1) =
    HMAC-SHA256(PRK, info || 0x01)."""
    h = hmac.HMAC(prk, hashes.SHA256())
    h.update(info + b"\x01")
    return h.finalize()


def _derive_keys(shared_secret: bytes, token: str):
    """PRK = HMAC-SHA256(key=token, msg=shared_secret) - i.e. the pre-shared
    token is HKDF's "salt", the ECDH output is the input keying material.
    Then two independent 32-byte AES keys, one per direction, so each side
    can maintain its own send/recv nonce counter without ever colliding."""
    prk = hmac.HMAC(token.encode("utf-8"), hashes.SHA256())
    prk.update(shared_secret)
    prk = prk.finalize()
    return _hkdf_one_block(prk, b"c2s"), _hkdf_one_block(prk, b"s2c")


def _encode_point(public_key) -> bytes:
    return public_key.public_bytes(
        serialization.Encoding.X962, serialization.PublicFormat.UncompressedPoint
    )


def _decode_point(data: bytes):
    return ec.EllipticCurvePublicKey.from_encoded_point(_CURVE, data)


def _send_line(sock, obj: dict):
    sock.sendall((json.dumps(obj) + "\n").encode("utf-8"))


def _recv_line(sock, timeout) -> bytes:
    sock.settimeout(timeout)
    buf = b""
    try:
        while b"\n" not in buf:
            chunk = sock.recv(4096)
            if not chunk:
                raise HandshakeError("connection closed during key exchange")
            buf += chunk
    except OSError as e:
        raise HandshakeError(f"read failed during key exchange: {e!r}") from e
    finally:
        sock.settimeout(None)
    return buf.split(b"\n", 1)[0]


class EncryptedConn:
    """Drop-in replacement for the raw-socket send/recv-line calls
    _StreamPhoneLink used to make directly - same blocking, line-oriented
    semantics, but everything after construction is AES-GCM encrypted.
    Building one performs the kex handshake, so construction blocks until
    that completes or fails.
    """

    def __init__(self, sock, token: str, is_server: bool):
        self._sock = sock
        self._send_counter = 0
        self._recv_counter = 0
        self._send_lock = threading.Lock()

        own_key = ec.generate_private_key(_CURVE)
        own_pub = _encode_point(own_key.public_key()).hex()

        if is_server:
            # Server speaks second: read the client's kex line, then reply.
            peer_line = _recv_line(sock, _KEX_TIMEOUT)
            peer_pub = _parse_kex(peer_line)
            _send_line(sock, {"type": "kex", "pub": own_pub})
        else:
            _send_line(sock, {"type": "kex", "pub": own_pub})
            peer_line = _recv_line(sock, _KEX_TIMEOUT)
            peer_pub = _parse_kex(peer_line)

        shared_secret = own_key.exchange(ec.ECDH(), _decode_point(bytes.fromhex(peer_pub)))
        c2s_key, s2c_key = _derive_keys(shared_secret, token)
        if is_server:
            self._recv_aead, self._send_aead = AESGCM(c2s_key), AESGCM(s2c_key)
        else:
            self._send_aead, self._recv_aead = AESGCM(c2s_key), AESGCM(s2c_key)

    def send_line(self, data: bytes):
        # Locked end-to-end (not just the counter increment): multiple
        # threads can call send() concurrently (each pending hook request
        # runs on its own thread), and interleaving the encrypt+write of two
        # messages would both scramble the framing and reuse a nonce out of
        # order - either one alone is bad, both together is worse.
        with self._send_lock:
            nonce = struct.pack(">Q", self._send_counter).rjust(12, b"\x00")
            self._send_counter += 1
            ct = self._send_aead.encrypt(nonce, data, None)
            self._sock.sendall(ct.hex().encode("ascii") + b"\n")

    def close(self):
        try:
            self._sock.close()
        except OSError:
            pass

    def recv_line(self, timeout=None):
        """Returns decrypted bytes, or None on a clean EOF (peer closed).
        socket.timeout/OSError propagate rather than collapsing to None, so
        a caller looping on this can tell "still connected, just quiet" from
        "connection is actually gone" apart - same distinction the raw
        socket recv() this replaces required its callers to make."""
        self._sock.settimeout(timeout)
        try:
            buf = b""
            while b"\n" not in buf:
                chunk = self._sock.recv(4096)
                if not chunk:
                    return None
                buf += chunk
        finally:
            self._sock.settimeout(None)
        line = buf.split(b"\n", 1)[0]
        nonce = struct.pack(">Q", self._recv_counter).rjust(12, b"\x00")
        self._recv_counter += 1
        try:
            return self._recv_aead.decrypt(nonce, bytes.fromhex(line.decode("ascii")), None)
        except Exception as e:
            raise HandshakeError(f"decrypt failed (wrong token, or tampering): {e!r}") from e


def _parse_kex(line: bytes) -> str:
    try:
        obj = json.loads(line)
    except json.JSONDecodeError as e:
        raise HandshakeError(f"bad kex JSON: {line!r}") from e
    if obj.get("type") != "kex" or "pub" not in obj:
        raise HandshakeError(f"expected kex message, got: {obj!r}")
    return obj["pub"]
